src/main/java/com/velox/sloan/cmo/workflows/planning/CustomFieldsRetriever.java

```diff
@@ -0,0 +1,464 @@
+
+                logError("Task does not have valid attachments or there are no records to display\n" +
+                        "There must be 'Sample' and 'PlanningStepProtocol1' records attached to this task to display custom fields.");
+            }
+            setPlanningStepValues(attachedSamples, attachedPlanningProtocols);
+
+        } catch (Exception e) {
+            clientCallback.displayError(String.format("Error while retrieving custom fields for samples:\n %s", e));
+            logError(String.format("Error while retrieving custom fields for samples:\n%s"), e);
+            return new PluginResult(false);
+        }
+        return new PluginResult(true);
+    }
+
+    /**
+     * The method checks if the active task contains Sample and ProtocolRecord attachments.
+     * @param attachedSamples
+     * @param attachedPlanninProtocols
+     * @return boolean value
+     */
+    private boolean hasValidAttachments(List<DataRecord> attachedSamples, List<DataRecord> attachedPlanninProtocols) {
+        return !(attachedSamples.isEmpty() && attachedPlanninProtocols.isEmpty());
+    }
+
+    /**
+     * This method looks for Sample that has a child record of type IndexBarcode. If not found on the first sample,
+     * the method keep looking upstream the parent hierarchy of the sample until a parent with a child record of IndexBarcode
+     * is found.
+     * @param sample
+     * @return String IndexId value for Sample if found, else returns "" with a warning.
+     * @throws NotFound
+     * @throws RemoteException
+     * @throws IoError
+     */
+    private String getSampleLibraryIndexId(DataRecord sample) throws NotFound, RemoteException, IoError {
+        DataRecord[] indexBarcodeRecord = sample.getChildrenOfType("IndexBarcode", user);
+        String indexId = "";
+        if (indexBarcodeRecord.length > 0) {
+            return indexBarcodeRecord[0].getStringVal("IndexId", user);
+        } else {
+            boolean found = false;
+            DataRecord startingSample = sample;
+            while (!found & startingSample != null) {
+                List<DataRecord> parentSamples = startingSample.getParentsOfType("Sample", user);
+                DataRecord parentSample = null;
+                if (!parentSamples.isEmpty()) {
+                    parentSample = parentSamples.get(0);
+                }
+                if (parentSample != null && parentSample.getChildrenOfType("IndexBarcode", user).length > 0
+                        && parentSample.getChildrenOfType("IndexBarcode", user)[0].getValue("IndexId", user) != null) {
+                    indexId = parentSample.getChildrenOfType("IndexBarcode", user)[0].getStringVal("IndexId", user);
+                    found = true;
+                } else {
+                    startingSample = parentSample;
+                }
+            }
+        }
+        return indexId;
```

Comment:
"indexId" may be empty if code reaches here, user Optional?

Reply:
An empty index ID is fine. I added a statement to show warning if Index ID is not found.